code/datums/world_topic.dm

```diff
@@ -1,246 +1,346 @@
-			warning("[WT] has keyword 'key'! Ignoring...")
-		else
-			.[keyword] = WT
+// VERSION
+
+// Update topic version whenever changes are made
+// The Version Number follows SemVer http://semver.org/
+#define TOPIC_VERSION_MAJOR		2	//	Major Version Number --> Increment when implementing breaking changes
+#define TOPIC_VERSION_MINOR		0	//	Minor Version Number --> Increment when adding features
+#define TOPIC_VERSION_PATCH		0	//	Patchlevel --> Increment when fixing bugs
 
 // DATUM
 
 /datum/world_topic
-	var/keyword
-	var/log = TRUE
-	var/key_valid
-	var/insecure_key = FALSE
-	var/require_comms_key = FALSE
-	var/permit_insecure = FALSE
-
-/datum/world_topic/proc/TryRun(list/input, addr)
-	key_valid = config && (CONFIG_GET(string/comms_key) == input["key"])
-	if(!key_valid && permit_insecure)
-		key_valid = config && (CONFIG_GET(string/comms_key_insecure) == input["key"])
-		insecure_key = key_valid
-	if(require_comms_key && !key_valid)
-		return "Bad Key"
-	if(insecure_key) // ignore the rate limiting if using true comms key
-		var/delta = world.time - GLOB.topic_cooldown
-		if(delta < CONFIG_GET(number/insecure_topic_cooldown))
-			return "Rate Limited"
-		GLOB.topic_cooldown = world.time
-	input -= "key"
-	. = Run(input, addr)
-	if(islist(.))
-		. = list2params(.)
-
-/datum/world_topic/proc/Run(list/input, addr)
-	CRASH("Run() not implemented for [type]!")
+	var/key // query key
+	var/anonymous = FALSE // can be used with anonymous authentication
+	var/list/required_params = list()
+	var/statuscode = null
+	var/response = null
+	var/data = null
+
+/datum/world_topic/proc/CheckParams(list/params)
+	var/list/missing_params = list()
+	var/errorcount = 0
+
+	for(var/param in required_params)
+		if(!params[param])
+			errorcount++
+			missing_params += param
+
+	if(errorcount)
+		statuscode = 400
+		response = "Bad Request - Missing parameters"
+		data = missing_params
+		return errorcount
+
+/datum/world_topic/proc/Run(list/input)
+	// Always returns true; actual details in statuscode, response and data variables
+	return TRUE
+
+// API INFO TOPICS
+
+/datum/world_topic/api_get_version
+	key = "api_get_version"
+	anonymous = TRUE
+
+/datum/world_topic/api_get_version/Run(list/input)
+	. = ..()
+	var/list/version = list()
+	var/versionstring = null
+
+	version["major"] = TOPIC_VERSION_MAJOR
+	version["minor"] = TOPIC_VERSION_MINOR
+	version["patch"] = TOPIC_VERSION_PATCH
+
+	versionstring = "[version["major"]].[version["minor"]].[version["patch"]]"
+
+	statuscode = 200
+	response = versionstring
+	data = version
+
+/datum/world_topic/api_get_authed_functions
+	key = "api_get_authed_functions"
+	anonymous = TRUE
+
+/datum/world_topic/api_get_authed_functions/Run(list/input)
+	. = ..()
+	var/list/functions = GLOB.topic_tokens[input["auth"]]
+	if(functions)
+		statuscode = 200
+		response = "Authorized functions retrieved"
+		data = functions
+	else
+		statuscode = 401
+		response = "Unauthorized - No functions found"
+		data = null
```

Comment:
I mean I don't see what's stopping people from just attempting every API endpoint anyways.

Reply:
Rate limits stop them, mostly. It at least adds a significant enough delay to make it difficult to go through the entire list at once.

Reply:
I guess I can remove/disable it if you'd like, but I really don't see how getting authorised endpoints is a security issue. If you leaked a token and someone wants to use it maliciously, removing this function will add max five minutes of work to just loop over the endpoints and sleep for 15 seconds to avoid rate limiting.

On the other hand, it's not really used for anything right now, since the only usecase within BYOND is already built into the handshake system.

HOWEVER, if you're interacting with an external tool, I think it's good to have a system to check what you can use, without having to actually poke the endpoints and see if you get 401 or not

Reply:
>  removing this function will add max five minutes of work to just loop over the endpoints and sleep for 15 seconds to avoid rate limiting.
That adds a pretty obvious pattern that can be observed. Even further, it'd be trivial to setup a monitor to watch for that kind of activity which can throw an alert for suspicious token activity. 

On one hand, I can see the value of this endpoint. On the other hand, I don't particularly like that it's an incredibly obvious attack vector.

Reply:
I am not so committed to this that I'm willing to die on the hill though. So, if the general consensus is that it'll probably be fine, I'm willing to resign from the debate and we can move forward with it still openly available.

Reply:
It'll probably be fine